src/AlgoliaSearch.js

```diff
@@ -1334,34 +1334,204 @@ AlgoliaSearch.prototype.Index.prototype = {
+   * to browse subsequent pages for this query. Use `index.browseNext(cursor)` when you want.
    *
-   * @param page Pagination parameter used to select the page to retrieve.
-   *             Page is zero-based and defaults to 0. Thus, to retrieve the 10th page you need to set page=9
-   * @param hitsPerPage: Pagination parameter used to select the number of hits per page. Defaults to 1000.
-   * @param callback the result callback called with two arguments:
-   *  error: null or Error('message'). If false, the content contains the error.
-   *  content: the server answer that contains the list of results.
+   * @param {string} query - The full text query
+   * @param {Object} [queryParameters] - Any search query parameter
+   * @param {Function} [callback] - The result callback called with two arguments
+   *   error: null or Error('message')
+   *   content: the server answer with user keys list
+   * @return {Promise|undefined} Returns a promise if no callback given
+   * @example
+   * index.browse('cool songs', {
+   *   tagFilters: 'public,comments',
+   *   hitsPerPage: 500
+   * }, callback);
+   * @see {@link https://www.algolia.com/doc/rest_api#Browse|Algolia REST API Documentation}
    */
-  browse: function(page, hitsPerPage, callback) {
+  // pre 3.5.0 usage, backward compatible
+  // browse: function(page, hitsPerPage, callback) {
+  browse: function(query, queryParameters, callback) {
+    var extend = require('extend');
+
     var indexObj = this;
 
-    if (arguments.length === 1 || typeof hitsPerPage === 'function') {
-      callback = hitsPerPage;
-      hitsPerPage = undefined;
-    }
+    var page;
+    var hitsPerPage;
 
-    var params = '?page=' + page;
-    if (!this.as._isUndefined(hitsPerPage)) {
-      params += '&hitsPerPage=' + hitsPerPage;
+    // we check variadic calls that are not the one defined
+    // .browse()/.browse(fn)
+    // => page = 0
+    if (arguments.length === 0 || arguments.length === 1 && typeof arguments[0] === 'function') {
+      page = 0;
+      callback = arguments[0];
+      query = undefined;
+    } else if (typeof arguments[0] === 'number') {
+      // .browse(2)/.browse(2, 10)/.browse(2, fn)/.browse(2, 10, fn)
+      page = arguments[0];
+      if (typeof arguments[1] === 'number') {
+        hitsPerPage = arguments[1];
+      } else if (typeof arguments[1] === 'function') {
+        callback = arguments[1];
+        hitsPerPage = undefined;
+      }
+      query = undefined;
+      queryParameters = undefined;
+    } else if (typeof arguments[0] === 'object') {
+      // .browse(queryParameters)/.browse(queryParameters, cb)
+      if (typeof arguments[1] === 'function') {
+        callback = arguments[1];
+      }
+      queryParameters = arguments[0];
+      query = undefined;
+    } else if (typeof arguments[0] === 'string' && typeof arguments[1] === 'function') {
+      // .browse(query, cb)
```

Comment:
It's already the case since the signature of the function is `function(query, queryParameters)`

I should also add that the order you do:

```js
      callback = arguments[0];
      query = undefined;
```

Is important since callback and query are references to arguments.

So if you do:

```js
      query = undefined;
      callback = arguments[0];
```

In every JS code this would work but here arguments[0] will be undefined as soon as query = undefined is executed.

DOH!JS!